### PR TITLE
First release 1.0.0

### PR DESCRIPTION
**Pull request by PayamEmami**
This is the first release of the workflow.

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
  - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
  - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/metaboigniter/tree/master/.github/CONTRIBUTING.md)
  - [ ] If necessary, also make a PR on the nf-core/metaboigniter _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint .`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (Leon-Bichmann) [inline .github/workflows/ci.yml:12]: If your pipeline has multiple modes of running, you can also add a few more tests with different parameter settings or modes  to this profile -> see for example the diaproteomics pipeline ci.yml

File: Dockerfile
Comment (Leon-Bichmann) [inline Dockerfile:1]: Did you accidentally delete your Dockerfile maybe? If you are not using DSL2 you still need it.

File: environment.yml
Comment (Leon-Bichmann) [inline environment.yml:1]: Same as your environment.yml how come you reomve it completely?

File: README.md
Comment (Leon-Bichmann) [inline README.md:80]: This part is still missing and can be quite helpful for users

File: README.md
Comment (Leon-Bichmann) [inline README.md:67]: if not available as preprint maybe delete the TODO lines

File: bin/scrape_software_versions.py
Comment (Leon-Bichmann) [inline bin/scrape_software_versions.py:8]: This part could be extended with other software you are using OpenMS for instance or others

File: conf/base.config
Comment (Leon-Bichmann) [inline conf/base.config:15]: Check with @ewels whether its ok to do it like this instead of environment.yml and dockerfile

File: conf/test_full.config
Comment (Leon-Bichmann) [inline conf/test_full.config:1]: see above

File: docs/output.md
Comment (Leon-Bichmann) [inline docs/output.md:9]: remove this line

File: environment.yml
Comment (Leon-Bichmann) [reply environment.yml:1]: ok I see you are trying to use the containers from the base.config

File: Dockerfile
Comment (Leon-Bichmann) [reply Dockerfile:1]: Ok I see you are trying to use the containers from the base.config

File: main.nf
Comment (Leon-Bichmann) [inline main.nf:458]: as mentioned above you could expand with other software such as OpenMS and co

File: main.nf
Comment (Leon-Bichmann) [inline main.nf:619]: I still find it a bit difficult to read these very long commands + corresponding parameters, maybe there is a way to structure this better with more newlines to improve readability

**Review by Leon-Bichmann — COMMENTED** [review]
Overall it looks good to me and it seems to run with the way you specify the containers. However sb from the core team would have to approve this way of managing the environment.

Apart from this small comments here and there, after an iteration of discussion and changes I can approve it for merging.

File: README.md
Comment (PayamEmami) [reply README.md:80]: Thanks for the comments. I will fix this.

File: conf/base.config
Comment (PayamEmami) [reply conf/base.config:15]: We have already been in contact several times. I think we came to the conclusion that making conda environment is difficult as we have many inconsistent packages. So it's okay like this. The solution would be to update the whole workflow which is on the agenda for the DSL 2

File: main.nf
Comment (PayamEmami) [reply main.nf:458]: Each process has it's own container. So i should either manually do that come up with a workaround. But it's a good suggestion

File: main.nf
Comment (PayamEmami) [reply main.nf:619]: Suggestion? I tried but it is still a mess

Comment (Leon-Bichmann) [inline main.nf:619]:
```suggestion
    /usr/local/bin/findPeaks.r input=\$PWD/!{mzMLFile} \
                                              output=\$PWD/!{mzMLFile.baseName}.rdata ppm=!{params.masstrace_ppm_pos_xcms} \
                                              peakwidthLow=!{params.peakwidthlow_quant_pos_xcms} \
```

File: main.nf
Comment (Leon-Bichmann) [inline main.nf:620]:
```suggestion
    peakwidthHigh=!{params.peakwidthhigh_quant_pos_xcms} \
    noise=!{params.noise_quant_pos_xcms} \
    polarity=positive realFileName=!{mzMLFile} \
    phenoFile=!{phenotype_file}  \
```

File: main.nf
Comment (Leon-Bichmann) [inline main.nf:621]:
```suggestion
   phenoDataColumn=!{params.phenodatacolumn_quant_pos} \
   sampleClass=!{params.sampleclass_quant_pos_xcms} \
   mzdiff=!{params.mzdiff_quant_pos_xcms} \
   snthresh=!{params.snthresh_quant_pos_xcms}  \
```

File: main.nf
Comment (Leon-Bichmann) [inline main.nf:622]:
```suggestion
  prefilter_l=!{params.prefilter_quant_pos_xcms} \
  prefilter_h=!{params.value_of_prefilter_quant_pos_xcms} \
  mzCenterFun=!{params.mzCenterFun_quant_pos_xcms} \
  integrate=!{params.integrate_quant_pos_xcms}  \
```

File: main.nf
Comment (Leon-Bichmann) [inline main.nf:623]:
```suggestion
 fitgauss=!{params.fitgauss_quant_pos_xcms} \
 methodXset=!{params.ipo_methodXset_pos} \
 methodRT=!{params.ipo_methodRT_pos} \
 noise_l=!{params.ipo_noise_l_pos} \
```

File: main.nf
Comment (PayamEmami) [reply main.nf:622]: Got. Will update the PR with the changes

File: main.nf
Comment (Leon-Bichmann) [inline main.nf:624]:
```suggestion
   noise_h=!{params.ipo_noise_h_pos} \
   prefilter_l_l=!{params.ipo_prefilter_l_l_pos} \
   prefilter_l_h=!{params.ipo_prefilter_l_h_pos} \
```

File: main.nf
Comment (Leon-Bichmann) [inline main.nf:625]:
```suggestion
  prefilter_h_l=!{params.ipo_prefilter_h_l_pos} \
  prefilter_h_h=!{params.ipo_prefilter_h_h_pos} \
  snthresh_l=!{params.ipo_snthresh_l_pos} \
  snthresh_h=!{params.ipo_snthresh_h_pos}  \
```

File: main.nf
Comment (Leon-Bichmann) [inline main.nf:626]:
```suggestion
 mzCenterFun=!{params.ipo_mzCenterFun_pos} \
 integrate=!{params.ipo_integrate_pos} \
 fitgauss=!{params.ipo_fitgauss_pos} \
 ipo_min_peakwidth_l=!{params.ipo_min_peakwidth_l_pos} \
```

File: main.nf
Comment (Leon-Bichmann) [inline main.nf:627]:
```suggestion
     ipo_min_peakwidth_h=!{params.ipo_min_peakwidth_l_pos} \
     ipo_max_peakwidth_l=!{params.ipo_max_peakwidth_l_pos} \
     ipo_max_peakwidth_h=!{params.ipo_max_peakwidth_h_pos} \
     ipo_ppm_l=!{params.ipo_ppm_l_pos} \
```

File: main.nf
Comment (Leon-Bichmann) [inline main.nf:628]:
```suggestion
   ipo_ppm_h=!{params.ipo_ppm_h_pos} \
   ipo_mzdiff_l=!{params.ipo_mzdiff_l_pos} \
   ipo_mzdiff_h=!{params.ipo_mzdiff_h_pos} \
   ipo_charge_camera=!{params.ipo_charge_camera_pos} \
   ipo_max_ppm_camera=!{params.ipo_max_ppm_camera_pos} \
```

File: main.nf
Comment (Leon-Bichmann) [inline main.nf:629]:
```suggestion
   ipo_inv=!{ipo_pos_localQ}  \
   ${filter_argument}
```

**Review by Leon-Bichmann — CHANGES_REQUESTED** [review]

File: conf/base.config
Comment (Leon-Bichmann) [reply conf/base.config:15]: ok then fine with me as well

File: README.md
Comment (ewels) [reply README.md:67]: This is the Zenodo DOI for the pipeline code itself, so should not be related to a preprint (but can add preprint / paper citations here in addition)

File: main.nf
Comment (ewels) [reply main.nf:458]: With DSL2 we are doing the version calls after the main step in the process, eg: [here](https://github.com/nf-core/modules/blob/983ba000c1d55297d8f32e29cac0a52229d1c757/software/rseqc/bamstat/main.nf#L37)

```nextflow
    output:
    tuple val(meta), path("*.my_output.txt"), emit: txt
    path  "*.version.txt"                  , emit: version

    script:
    def software = getSoftwareName(task.process)
    """
    my_command.py -i $bam $options.args > ${prefix}.my_output.txt
    my_command.py --version > ${software}.version.txt
    """
```

I think that it would be a good approach to do the same here, then scoop all of these channels up into the software_versions process. This way you get the versions and you're already part-way towards DSL2 style..

File: main.nf
Comment (ewels) [reply main.nf:619]: I think you might need double-slashes for it to get through Nextflow parsing nicely (otherwise the final command in the output just has huge chunks of whitespace, that makes it difficult to read in the pipeline report etc).

Comment (ewels) [reply main.nf:619]: yeah, as @Leon-Bichmann line break bash backslashes are the way to go 👍🏻 eg. his example below or also [this one](https://github.com/nf-core/modules/blob/983ba000c1d55297d8f32e29cac0a52229d1c757/software/star/align/main.nf#L48-L56)

**Comment by ewels** [comment]
Might be worth taking another look at my PR @PayamEmami : https://github.com/nf-core/metaboigniter/pull/8 as I already tackled a few of the suggestions from @Leon-Bichmann (eg. breaking commands over lines). But I guess that PR is now behind your template sync so probably worth closing and ignoring the actual changes I did as merging would be complex.

My general points still stand though I think:

* Use `"""` instead of `'''` for script blocks unless there's a very good reason
  * Usually `'''` is only needed when using a command that has loads of environment variables with `$` that you don't want to escape
* Use simpler `$params.foo` instead of `!{params.food}` unless there's a good reason
  * Same as above. Also most code editor syntax highlighters ignore `!{params.food}` but flag `$params.foo` which makes the latter much easier to read.
* Split commands to have one argument per line to make them easier to read

You [replied](https://github.com/nf-core/metaboigniter/pull/8#issuecomment-788730473) that:

> I see your points. Unfortunately, using `"""` in some instances will fail. That's because I used `'''`. The same goes for the parameters. If I use `'''`, then I will have to use `!{}` ...

I don't really understand what you mean. Yes, `"""` goes with `$` and `'''` goes with `!{}`. But I'm suggesting that you try to use the former wherever you can. Basically the only time you need the latter is when you are making extensive use of bash variables in the command and I don't really see this anywhere in your pipeline. The former is a lot more standard and generally recommended as it works better with code syntax highlighting and lots of other small things.

File: conf/base.config
Comment (ewels) [reply conf/base.config:15]: I think it would be good to have these under the `nfcore` DockerHub account if possible. Mostly because we have `nfcore` registered as an open-source community there so the account is not affected by pull limits and auto-archive / deletion policies. Without this I'd worry that the images might get automatically removed in a year or two. For example:

```suggestion
withLabel: openms      { container = "nfcore/metaboigniter-openms:v2.4.0" }
```

It would also be good to have the `Dockerfile` scripts in the pipeline for clarity (eg. in a subfolder).

Happy to help with both of these points.

File: .github/workflows/awsfulltest.yml
Comment (PayamEmami) [reply .github/workflows/awsfulltest.yml]: It has now been added.

File: .github/workflows/ci.yml
Comment (PayamEmami) [reply .github/workflows/ci.yml:12]: Agree but for now, it's enough with one mode. The full test will do the rest for more advanced options. Thanks

File: Dockerfile
Comment (PayamEmami) [reply Dockerfile:1]: Now there is a single docker container.

File: environment.yml
Comment (PayamEmami) [reply environment.yml:1]: See above

File: README.md
Comment (PayamEmami) [reply README.md]: Thanks. Added!

File: conf/base.config
Comment (PayamEmami) [reply conf/base.config]: Fixed

File: conf/test_full.config
Comment (PayamEmami) [reply conf/test_full.config:1]: Fixed

File: docs/output.md
Comment (PayamEmami) [reply docs/output.md]: Fixed

File: main.nf
Comment (PayamEmami) [reply main.nf:476]: Fixed using a single container.

File: main.nf
Comment (PayamEmami) [reply main.nf]: All have been fixed. Thanks, guys for the comment.

Comment (PayamEmami) [reply main.nf]: Fixed

File: main.nf
Comment (PayamEmami) [reply main.nf]: fixed

File: main.nf
Comment (PayamEmami) [reply main.nf]: fixed

**Comment by PayamEmami** [comment]
I fixed this as you suggested. Thanks


> Might be worth taking another look at my PR @PayamEmami : #8 as I already tackled a few of the suggestions from @Leon-Bichmann (eg. breaking commands over lines). But I guess that PR is now behind your template sync so probably worth closing and ignoring the actual changes I did as merging would be complex.
> 
> My general points still stand though I think:
> 
> * Use `"""` instead of `'''` for script blocks unless there's a very good reason
>   
>   * Usually `'''` is only needed when using a command that has loads of environment variables with `$` that you don't want to escape
> * Use simpler `$params.foo` instead of `!{params.food}` unless there's a good reason
>   
>   * Same as above. Also most code editor syntax highlighters ignore `!{params.food}` but flag `$params.foo` which makes the latter much easier to read.
> * Split commands to have one argument per line to make them easier to read
> 
> You [replied](https://github.com/nf-core/metaboigniter/pull/8#issuecomment-788730473) that:
> 
> > I see your points. Unfortunately, using `"""` in some instances will fail. That's because I used `'''`. The same goes for the parameters. If I use `'''`, then I will have to use `!{}` ...
> 
> I don't really understand what you mean. Yes, `"""` goes with `$` and `'''` goes with `!{}`. But I'm suggesting that you try to use the former wherever you can. Basically the only time you need the latter is when you are making extensive use of bash variables in the command and I don't really see this anywhere in your pipeline. The former is a lot more standard and generally recommended as it works better with code syntax highlighting and lots of other small things.

File: Dockerfile
Comment (Leon-Bichmann) [reply Dockerfile:1]: ok fine with me, if @Phil agrees to have it this way

File: conf/test_full.config
Comment (Leon-Bichmann) [reply conf/test_full.config:1]: ok so those are the same test files as for the small test right?

In principle it is fine with me however you could for the full_test as it will be running on AWS also choose a different much larger test-dataset. Sth. from a public metabolomics repository that would be to large to run it in the github actions

File: nextflow_schema.json
Comment (Leon-Bichmann) [inline nextflow_schema.json:2895]: I think what @ewels meant before in the slack chat was, that these headings eg. `## Centroiding` within the help_text entries confuse the website parser. I would remove the hashtags. Check also in other help_texts

Comment (Leon-Bichmann) [inline nextflow_schema.json:2895]:
```suggestion
            "help_text": "Centroiding\n\nWe recommend inputting already centroided files. You can achieve this at the conversion steps in msconvert. However, if your data is not centroided, you can let the workflow doing that for you. We use OpenMS \"PeakPickHiRes\" tool to perform that. Set the following parameter to *true* to perform the centroiding:\n\n```nextflow\n    need_centroiding\n```\n\nPlease be aware that setting need_centroiding=true will do centroiding on all of your data including both ionization, identification etc.\n\nIf you want non-defualt values, to control the parameters centroiding you can edit *openms_peak_picker_ini_pos.ini* and *openms_peak_picker_ini_neg.ini* files located under *metaboigniter/assets/openms*. The description of the parameters can be found on [OpenMS website](https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_PeakPickerHiRes.html).\n\nIf you choose to perform the quantification using OpenMS, you should consider changing the parameters for OpenMS only. You can edit *openms_feature_finder_metabo_ini_pos.ini* and *openms_feature_finder_metabo_ini_neg.ini* files located under *metaboigniter/assets/openms*. The description of the parameters can be found on [OpenMS website](https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_FeatureFinderMetabo.html). **When tuning the OpenMS parameters make sure that \"report_convex_hulls\" is set to true**\n\n## Ionization\n\nSet what type of ionization you have. You can either set to 'pos' (only positive), 'neg' (only negative), 'both' (both positive and negative):\n\n```nextflow\n      type_of_ionization\n```\n\n## Identification Specific Parameters\n\nMetaboIGNITER has a specific design for performing identification. After performing quantification and annotation of MS1 data, the results will be fed into the identification sub-pipeline this part of the workflow, will first extract the MS2 information from the identification mzML files and map them to the quantification data. If a parent ion was matched against a feature that was successfully annotated, we then estimate the neutral mass for that parent ion. By doing that, the number of searches needed for identification will significantly decrease. The rest of the ions that were matched against features without annotation will be searched using different adduct rules and charges. The resulting scores from metabolite spectrum matches will be transformed into posterior error probability and re-matched to the features at the later step.\n\nCurrently, MetaboIGNITER supports two types of identification. One is based on in-silico (database) approach and the other is based on internal library.\n\nIf you would like to perform identification, please set the following parameters to true:\n\n```nextflow\n    perform_identification=true\n```\n\nPlease also remember that we only perform identification for the ionization modes that you have quantification data for.\n\n## In-silico identification in MetaboIGNITER\n\nCurrently, MetaboIGNITER supports three search engines, [MetFrag](https://ipb-halle.github.io/MetFrag/), [CSI:FingerID](https://bio.informatik.uni-jena.de/software/sirius/), and [CFM-ID](https://cfmid.wishartlab.com/) for performing identification. These engines share some global parameters but also each of these will need specific set of parameters. The user can select multiple search engines to do the identification. In the case of multiple search engines, the workflow will have multiple final output, one for each search engine and ionization. Please set the parameter to true for the desired search engine:\n\n```nextflow\n    perform_identification_metfrag=false \n    perform_identification_csifingerid=false\n    perform_identification_cfmid=false \n```\n\nCharacterize your own library\n\nThis part of the workflow is used to create and characterize in-house library. This is how it works: we assume that the library consists of one or more mzML files, each containing a number of compounds. A possible scenario is when the users have several standard metabolites that can have overlapping masses with unknown retention times. The standards with overlapping masses can be run separately using MS, resulting in different runs. MetaboIGNITER will help you to characterize this type of internal libraries. You will need to construct the Characterization file (see below) that shows which standards are present in which mzML file. The workflow will then do mass trace detection, MS2 extraction and mapping of parent ions to mass traces. Doing so will result in finding the retention time and empirical m/z of each standard. This will then be used to create identification parameters and search the biological MS2 files.\n\nSet the library parameter to true if you would like to perform library search:\n\n```nextflow\n    perform_identification_internal_library=true\n```\n\n## Output files\nThere will be a lot of different files generated by the workflow, if you are interested only in the final output part and identification information, please set this flag to false\n\n```nextflow\n    publishDir_intermediate\n```"
```

**Review by Leon-Bichmann — COMMENTED** [review]
apart from these minor things it looks good to me.

File: nextflow_schema.json
Comment (PayamEmami) [reply nextflow_schema.json:2895]: Aaah. Then i completely misunderstood  it! Will fix. More comments?

File: conf/test_full.config
Comment (PayamEmami) [reply conf/test_full.config:1]: Great point. I was actually thinking about this and we do have a big dataset. But running the workflow with a large dataset will take weeks to finish.

File: nextflow_schema.json
Comment (PayamEmami) [reply nextflow_schema.json]: The problem appears to be fixed. Thanks for the suggestion!

**Review by Leon-Bichmann — APPROVED** [review]
Ok, if the tests pass this is ready for merging now from my side !

**Comment by ewels** [comment]
I'm looking into fixing the whitespace and some code style issues currently. PR incoming soon.

**Review by ewels — APPROVED** [review]
Once the other open PRs are merged, I think I will be happy with this. Some things  that would be nice to improve:

* Documentation, especially the lack of output documentation
* Software packaging, but we've talked about that before and I'm a lot happier with it now than I was before 👍🏻 
* Code duplication in the main pipeline, but I hope that this can be tackled in future work to convert the pipeline to DSL2.

File: main.nf
Comment (ewels) [reply main.nf:48]: Done in https://github.com/nf-core/metaboigniter/pull/29

File: main.nf
Comment (ewels) [reply main.nf:25]: Done in https://github.com/nf-core/metaboigniter/pull/29

File: CHANGELOG.md
Comment (ewels) [reply CHANGELOG.md:3]:
```suggestion
## [v1.0.0](https://github.com/nf-core/methylseq/releases/tag/1.0.0) - 2021-05-08
```

**Pull request merged**
